src/glyf.cc

```diff
@@ -302,9 +313,68 @@ bool OpenTypeGLYF::Parse(const uint8_t *data, size_t length) {
         return Error("Failed to parse glyph %d", i);
       }
     } else {
-      if (!ParseCompositeGlyph(glyph)) {
+
+      ComponentPointCount component_point_count;
+      component_point_count.accumulated_component_points = 0;
+      component_point_count.gid_stack.clear();
+
+      if (!ParseCompositeGlyph(glyph, &component_point_count)) {
         return Error("Failed to parse glyph %d", i);
       }
+
+      // Check maxComponentDepth and validate
+      // maxComponentPoints. ParseCompositeGlyph placed the first set of
+      // component glyphs on the component_point_count.gid_stack, which we start
+      // to process below. If a nested glyph is in turn a component glyph,
+      // additional glyphs are placed on the stack.
+      while (component_point_count.gid_stack.size()) {
+        GidAtLevel stack_top_gid = component_point_count.gid_stack.back();
+        component_point_count.gid_stack.pop_back();
+
+        const unsigned gly_offset = offsets[stack_top_gid.gid];
+        const unsigned gly_length = offsets[stack_top_gid.gid + 1] -
+                                    offsets[stack_top_gid.gid];
+        if (!gly_length) {
+          continue;
+        }
+
+        if (gly_offset >= length) {
+          return Error("Glyph %d offset %d too high %ld",
+                       i,
+                       gly_offset,
+                       length);
+        }
+        if (gly_offset + gly_length < gly_offset) {
+          return Error("Glyph %d length (%d < 0)!", i, gly_length);
+        }
+        if (gly_offset + gly_length > length) {
+          return Error("Glyph %d length %d too high", i, gly_length);
+        }
+
+        Buffer points_count_glyph(data + gly_offset, gly_length);
```

Comment:
I feel uneasy about the duplicated code (from `gly_offset` to here, may be it should be factored out (`GetGlyphData()` or something)?

Reply:
Factored out glyph buffer retrieval out into new function `GetGlyphBufferSection()`.